sdk/go/auto/cmd.go

```diff
@@ -17,17 +17,213 @@ package auto
+// the installation root or to validate that the version meets a minimum
+// requirement. By default the pulumi binary found in $PATH is used.
+func NewPulumiCommand(opts ...PulumiCommandOption) (PulumiCommand, error) {
+	pOpts := &pulumiCommandOptions{}
+	for _, opt := range opts {
+		opt.applyPulumiCommandOption(pOpts)
+	}
+
+	command := "pulumi"
+	if pOpts.root != "" {
+		command = path.Join(pOpts.root, "bin", "pulumi")
+	}
+
+	cmd := exec.Command(command, "version")
+	out, err := cmd.Output()
+	if err != nil {
+		return pulumiCommand{}, fmt.Errorf("failed to run `pulumi version`: %w", err)
+	}
+	currentVersion := strings.TrimSpace(string(out))
+	min := minimumVersion
+	if pOpts.version.GT(min) {
+		min = pOpts.version
+	}
+	version, err := parseAndValidatePulumiVersion(min, currentVersion, pOpts.skipVersionCheck)
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+
+	return pulumiCommand{
+		version: version,
+		command: command,
+	}, nil
+}
+
+// Install downloads and installs the Pulumi CLI. By default the CLI version
+// matching the current SDK release is installed in
+// $HOME/.pulumi/versions/$VERSION. Use the PulumiCommandOption `Root` to
+// specify a different directory, and `Version` to install a custom version.
+func Install(ctx context.Context, opts ...PulumiCommandOption) (PulumiCommand,
+	error,
+) {
+	pOpts := &pulumiCommandOptions{}
+	for _, opt := range opts {
+		opt.applyPulumiCommandOption(pOpts)
+	}
+	err := pOpts.applyDefaults()
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+	fullOpts := []PulumiCommandOption{Root(pOpts.root), Version(pOpts.version), skipVersionCheck(pOpts.skipVersionCheck)}
+	pulumi, err := NewPulumiCommand(fullOpts...)
+	if err == nil {
+		// Found an installation and it satisfies the version requirement
+		return pulumi, nil
+	}
+	if runtime.GOOS == "windows" {
+		if err := installWindows(ctx, pOpts.version, pOpts.root); err != nil {
+			return pulumiCommand{}, err
+		}
+	} else {
+		if err := installPosix(ctx, pOpts.version, pOpts.root); err != nil {
+			return pulumiCommand{}, err
+		}
+	}
+	return NewPulumiCommand(fullOpts...)
+}
+
+func installWindows(ctx context.Context, version semver.Version, root string) error {
+	url := "https://get.pulumi.com/install.ps1"
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := httputil.DoWithRetry(req, http.DefaultClient)
+	if err != nil {
+		return fmt.Errorf("failed to download %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+	scriptData, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response from %s: %w", url, err)
+	}
+	tmp, err := os.CreateTemp("", "install-*.ps1")
+	if err != nil {
+		return err
+	}
+	scriptPath := tmp.Name()
+	defer os.Remove(scriptPath)
+	err = os.WriteFile(scriptPath, scriptData, 0o600)
+	if err != nil {
+		return err
+	}
+	if err = tmp.Close(); err != nil {
+		return err
+	}
```

Comment:
This seems pretty much the same code as we have for installPosix.  Is it worth cleaning this up?